File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
# import bisect
def binary_search(tails, num):
    low, high = 0, len(tails)
    while low<high:
        mid = (low+high)//2
        if tails[mid]<num:
            low = mid +1
        else:
            high = mid
    return low


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        temp = []
        for num in nums:
            idx = binary_search(temp, num)
            if idx<len(temp):
                temp[idx] = num
            else:
                temp.append(num)
        return len(temp)

        

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.go
package main
import (
    "fmt"
    "sort"
)
func lengthOfLIS(nums []int) int {
  if len(nums)==0{
    return 0
  }

  var tails []int
  var tails_indices []int

  perent := make([]int, len(nums))
  for i := range perent{
    perent[i] = -1
  }

  for i , num := range nums {
    idx := sort.SearchInts(tails, num)

    if idx == len(tails){
        tails = append(tails, num)
        tails_indices = append(tails_indices, i)
    } else{
        tails[idx] = num
        tails_indices[idx] = i
    }

    if idx>0{
        perent[i] = tails_indices[idx-1]
    }

  }
  var result []int
  current := tails_indices[len(tails_indices)-1]
  for current != -1{
    result = append(result, nums[current])
    current = perent[current]
  }

  return len(result)
}